test/common/config/delta_subscription_state_test.cc

```diff
@@ -388,6 +403,58 @@ TEST_F(DeltaSubscriptionStateTest, AddedAndRemoved) {
             ack.error_detail_.message());
 }
 
+TEST_F(DeltaSubscriptionStateTest, ResourceTTL) {
+  Event::SimulatedTimeSystem time_system;
+  time_system.setSystemTime(std::chrono::milliseconds(0));
+
+  auto create_resource_with_ttl = [](absl::optional<std::chrono::seconds> ttl_s) {
+    Protobuf::RepeatedPtrField<envoy::service::discovery::v3::Resource> added_resources;
+    auto* resource = added_resources.Add();
+    resource->set_name("name1");
+    resource->set_version("version1A");
+
+    if (ttl_s) {
+      ProtobufWkt::Duration ttl;
+      ttl.set_seconds(ttl_s->count());
+      resource->mutable_ttl()->CopyFrom(ttl);
+    }
+
+    return added_resources;
+  };
+
+  {
+    EXPECT_CALL(*timer_, enabled());
+    EXPECT_CALL(*timer_, enableTimer(_, _));
+    deliverDiscoveryResponse(create_resource_with_ttl(std::chrono::seconds(1)), {}, "debug1",
+                             "nonce1");
+  }
+
+  {
+    // Increase the TTL.
+    EXPECT_CALL(*timer_, enabled());
+    EXPECT_CALL(*timer_, enableTimer(_, _));
+    deliverDiscoveryResponse(create_resource_with_ttl(std::chrono::seconds(2)), {}, "debug1",
+                             "nonce1");
+  }
+
+  // Remove the TTL.
+  EXPECT_CALL(*timer_, disableTimer());
+  deliverDiscoveryResponse(create_resource_with_ttl(absl::nullopt), {}, "debug1", "nonce1");
+
+  // Add back the TTL.
+  EXPECT_CALL(*timer_, enabled());
+  EXPECT_CALL(*timer_, enableTimer(_, _));
+  deliverDiscoveryResponse(create_resource_with_ttl(std::chrono::seconds(2)), {}, "debug1",
```

Comment:
Other than timer toggling, are there stronger properties that can be expected/asserted, e.g. as we timeout?